### ISSUE TITLE
PulseEmitter is difficult to test with

### ISSUE THREAD

**Issue by apoydence** [cloudfoundry/go-loggregator#20]
### tl;dr
The output types of `NewCounterMetric(...)` and `NewGaugeMetric(...)`  (`*CounterMetric` and `*GaugeMetric` respectively) do not offer a user a straightforward path to create a spy.

### Problem
While a `PulseEmitter` could be injected, the `CounterMetric` and `GaugeMetric` force the user to submit an envelope, and then query the underlying contents of the envelope. This implies the user has to have intimate knowledge about Loggregator envelopes, which goes against the goal of go-loggregator.

### Proposed Solution
Update `NewCounterMetric(...)` and `NewGaugeMetric(...)` to return interface types:
```go
type Gauge interface {
  Set(int64)
}
```
and
```go
type Counter interface {
  Increment(uint64)
}
```

This would allow the entire `PulseEmitter` to be replaced, in-full, with a spy/mock. It would also enable the deletion of `GetDelta()` from `CounterMetric`. This method is only useful from a test perspective and could easily be "cleaned" away and break test compatibility.

**Comment by jasonkeene** [comment]
Via @bradylove:

> PRs welcome :trollface:
